### PR TITLE
KAFKA-4547: Avoid unnecessary offset commit that could lead to an invalid offset position if partition is paused

### PR DESCRIPTION
**Pull request by vahidhashemian**

### CODE REVIEW

**Comment by vahidhashemian** [comment]
@hachikuji I submitted this PR without test(s) in case it needs to be included in 0.10.2.0. I'll work on creating some tests after I'm done with the PR for KIP-88.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 1]
```diff
@@ -334,6 +334,13 @@ public OffsetResetStrategy resetStrategy(TopicPartition partition) {
         return assignedState(partition).resetStrategy;
     }
 
+    public boolean haveAllFetchPositions(Collection<TopicPartition> partitions) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Can we use "has" for consistency?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -1569,8 +1569,8 @@ private void updateFetchPositions(Set<TopicPartition> partitions) {
         // the user is manually assigning partitions and managing their own offsets).
         fetcher.resetOffsetsIfNeeded(partitions);
 
-        if (!subscriptions.hasAllFetchPositions()) {
-            // if we still don't have offsets for all partitions, then we should either seek
+        if (!subscriptions.haveAllFetchPositions(partitions)) {
+            // if we still don't have offsets for given partitions, then we should either seek
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: "for _the_ given partitions"?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java (hunk 1 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Sure, I'll update it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 3]
```diff
@@ -334,6 +334,13 @@ public OffsetResetStrategy resetStrategy(TopicPartition partition) {
...
     }
 
+    public boolean hasAllFetchPositions(Collection<TopicPartition> partitions) {
+        for (TopicPartition partition : partitions)
+            if (!hasValidPosition(partition))
+                return false;
+        return true;
+    }
+
     public boolean hasAllFetchPositions() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Seems we could simplify this by delegating to the other method.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Or maybe we don't even need this method anymore (haven't checked)

Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: It's being used in a couple of place: [here](https://github.com/apache/kafka/blob/d1733953742236e0430aff4e502bd8ee7c642591/clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java#L51) and [here](https://github.com/apache/kafka/blob/d1733953742236e0430aff4e502bd8ee7c642591/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L1043).

Even though they can be modified to leverage the new method, I'd prefer to keep it and just modify its implementation to delegate to the other one as you suggested. What do you think?

Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: I submitted an update that refactors the original method. I'd be happy to take your second suggestion too if you think that's a better way to go. Thanks.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: If it's used, we can keep it. I wasn't sure, so thought I'd mention it.

**Comment by ewencp** [comment]
@hachikuji @vahidhashemian Please note that I've updated the affected versions to 0.10.0.2 and 0.10.1.0, 0.10.1.1. When we merge a fix, we want to make sure it makes it onto branches 0.10.0, 0.10.1, 0.10.2, and trunk.

**Comment by hachikuji** [comment]
@vahidhashemian The fix looks good, but can you add a couple test cases?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 4]
```diff
@@ -1094,6 +1094,60 @@ public void testManualAssignmentChangeWithAutoCommitDisabled() {
         consumer.close();
     }
 
+    @Test
+    public void testOffsetOfPausedPartitions() {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1099]: Thanks for adding the test. Would be good to have a more focused test in `FetcherTest` as well for the change there.

Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1099]: Sure, I'll try to add that too. Thanks for the quick feedback.

**Comment by ijuma** [comment]
@ewencp does it affect 0.10.0.2 as well? The reporter suggested that 0.10.0.1 worked fine while 0.10.1.0 did not, so I was wondering if you had additional information than what was in the JIRA.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java [hunk 5]
```diff
@@ -545,6 +545,21 @@ public void testUpdateFetchPositionDisconnect() {
     }
 
     @Test
+    public void testUpdateFetchPositionOfPausedPartitions() {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java]: Could we have one more test case similar to this, but in which the paused partition already has a position? This verifies that `updateFetchPositions` does not overwrite the current position.

Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java]: Sure. I'll add one more. Thanks.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 6]
```diff
@@ -1094,6 +1094,60 @@ public void testManualAssignmentChangeWithAutoCommitDisabled() {
...
+
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
+                rebalanceTimeoutMs, sessionTimeoutMs, heartbeatIntervalMs, true, autoCommitIntervalMs);
+
+        // lookup coordinator
+        client.prepareResponseFrom(new GroupCoordinatorResponse(Errors.NONE.code(), node), node);
+        Node coordinator = new Node(Integer.MAX_VALUE - node.id(), node.host(), node.port());
+
+        // manual assignment
+        List<TopicPartition> partitions = Arrays.asList(tp0, tp1);
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: nit: if you use `Utils.mkSet`, then you won't need the conversion below.

**Comment by hachikuji** [comment]
LGTM. I think @ijuma is right that this only needs to go into 0.10.1 and 0.10.2.

**Comment by hachikuji** [comment]
@vahidhashemian Merged to trunk and 0.10.2. Unfortunately, it doesn't merge cleanly onto 0.10.1. Would you mind submitting a separate PR to that branch?

**Comment by vahidhashemian** [comment]
@hachikuji Definitely.

**Pull request closed without merging**